heat_pump_adoption_modelling/pipeline/supervised_model/data_aggregation.py

```diff
@@ -0,0 +1,429 @@
+    categorical_features = [
+        feature
+        for feature in df.columns
+        if (feature not in ordinal_features)
+        and (feature not in numeric_features)
+        and (feature not in unaltered_features)
+    ]
+
+    # Aggreate categorical features
+    cat_agglomerated = aggreate_categorical_features(
+        df[categorical_features], categorical_features, agglo_feature=postcode_level
+    )
+
+    # Get the features indicating the most frequent values
+    most_frequent = [col for col in cat_agglomerated.columns if "MOST_FREQUENT" in col]
+
+    # One-hot encode categorical features
+    cat_agglomerated = feature_encoding.one_hot_encoding(
+        cat_agglomerated, most_frequent, verbose=True
+    )
+
+    # Concatenate the processed categorical, numeric and unaltered features
+    agglomerated_df = pd.concat(
+        [
+            cat_agglomerated,
+            num_agglomerated.drop(columns=[postcode_level]),
+            df[unaltered_features],
+        ],
+        axis=1,
+    )
+
+    return agglomerated_df
+
+
+def get_target_variables(
+    df, source_year, target_year, postcode_level, normalise="total"
+):
+    """Get the target variables (current and future heat pump coverage and growth)
+    based on source and target year data.
+
+    Parameters
+    ----------
+    df: pandas.Dataframe
+        Dataframe with all year's data, used for normalising.
+
+    source_year: pandas.Dataframe
+        Dataframe with data for source year.
+
+    target_year: pandas.Dataframe
+        Dataframe with data for target year.
+
+    postcode_level: str
+        Postcode level to group by.
+
+    normalise: str, default='total'
+        How to normalise the number of properties per postcode.
+        'total' : normalise by total number of properties
+        'target' : normalise by number of properties at target year
+
+    Return
+    ---------
+    target_var_df : pandas.DataFrame
+        Target variable dataframe with heat pump coverage and growth information."""
+
+    # Get the number of installed heat pumps per group (e.g. postcode level)
+    n_hp_installed_source = get_feature_count_grouped(
+        source_year, "HP_INSTALLED", postcode_level, name="# HP at Time t"
+    )
+    n_hp_installed_target = get_feature_count_grouped(
+        target_year, "HP_INSTALLED", postcode_level, name="# HP at Time t+1"
+    )
+    n_hp_installed_total = get_feature_count_grouped(
+        df, "HP_INSTALLED", postcode_level, name="Total # HP"
+    )
+
+    # Select basis for normalisation
+    if normalise == "total":
+        total_basis = df
+    elif normalise == "target":
+        total_basis = target_year
+    else:
+        raise IOError(
+            "Normalisation type '{}' is not defined. Please use 'total' or 'target' instead.".format(
+                normalise
+            )
+        )
+
+    # Get the number of properties for each postcode
+    n_prop_total = (
```

Comment:
Yes,  only 1 record per property should be considered. df does not have to be deduplicated, this will be taken care of. I rearranged some things to make it clearer.